finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4StreamTransport.scala

```diff
@@ -435,29 +441,40 @@ private[h2] trait Netty4StreamTransport[SendMsg <: Message, RecvMsg <: Message]
             else recv(data)
 
           case Open(remote@RemoteStreaming()) =>
-            if (data.isEndStream) {
-              if (stateRef.compareAndSet(state, remote.toRemoteClosed)) {
-                if (recvFrame(toFrame(data), remote)) true
-                else throw new IllegalStateException("stream queue closed prematurely")
-              } else recv(data)
-            } else {
+            if (stateRef.compareAndSet(state, remote.toRemoteClosed)) {
               if (recvFrame(toFrame(data), remote)) true
-              else recv(data)
-            }
+              else throw new IllegalStateException("stream queue closed prematurely")
+            } else recv(data)
 
           case LocalClosed(remote@RemoteStreaming()) =>
-            if (data.isEndStream) {
-              if (stateRef.compareAndSet(state, Closed(Reset.NoError))) {
-                if (recvFrame(toFrame(data), remote)) {
-                  remote.close()
-                  resetP.setDone()
-                  true
-                } else throw new IllegalStateException("stream queue closed prematurely")
-              } else recv(data)
-            } else {
-              if (recvFrame(toFrame(data), remote)) true
-              else recv(data)
-            }
+            if (stateRef.compareAndSet(state, Closed(Reset.NoError))) {
+              if (recvFrame(toFrame(data), remote)) {
+                remote.close()
+                resetP.setDone()
+                true
+              } else throw new IllegalStateException("stream queue closed prematurely")
+            } else recv(data)
+        }
+
+      case data: Http2DataFrame =>
+        state match {
+          case Closed(_) => false
+
+          case state@RemoteClosed() =>
+            if (resetFromLocal(state, Reset.Closed)) false
+            else recv(data)
+
+          case RemoteOpen(remote@RemotePending()) =>
+            if (resetFromLocal(remote, Reset.InternalError)) false
+            else recv(data)
+
+          case Open(remote@RemoteStreaming()) =>
+            if (recvFrame(toFrame(data), remote)) true
+            else recv(data)
+
+          case LocalClosed(remote@RemoteStreaming()) =>
+            if (recvFrame(toFrame(data), remote)) true
+            else recv(data)
```

Comment:
This is just a refactor. I kept getting confused that we process `Http2HeadersFrame.isEndStream` and `Http2HeadersFrame` in separate cases, but we process all `Http2DataFrame` in the same case. Breaking data frames out into two separate cases feels more consistent with how we're handling header frames.